File: algorithms/greedy_alg/greedy_alg.py
states_needed = set(['mt', 'wa', 'or', 'id', 'nv', 'ut', 'ca', 'az'])

stations = {}
stations['kone'] = set(['id', 'nv', 'ut'])
stations['ktwo'] = set(['wa', 'id', 'mt'])
stations['kthree'] = set(['or', 'nv', 'ca'])
stations['kfour'] = set(['nv', 'ut'])
stations['kfive'] = set(['ca', 'az'])

def coverSet(states_needed, stations):
    final_staions = set()

    while states_needed: 
        best_station = None
        states_covered = set()

        for station, states_for_station in stations.items():
            covered = states_needed & states_for_station
            if len(covered) > len(states_covered):
                best_station = station
                states_covered = covered

        final_staions.add(best_station)

        states_needed -= states_covered

    return final_staions

print(coverSet(states_needed, stations))

File: algorithms/greedy_alg/greedy_alg.js
function union(a, b) {
    return new Set([...a, ...b]);
}

function intersection(a, b) {
    return new Set([...a].filter((x) => b.has(x)));
}

function difference(a, b) {
    return new Set([...a].filter((x) => !b.has(x)));
}

let statesNeeded = new Set(["mt", "wa", "or", "id", "nv", "ut", "ca", "az"]);

const stations = {};
stations["kone"] = new Set(["id", "nv", "ut"]);
stations["ktwo"] = new Set(["wa", "id", "mt"]);
stations["kthree"] = new Set(["or", "nv", "ca"]);
stations["kfour"] = new Set(["nv", "ut"]);
stations["kfive"] = new Set(["ca", "az"]);

function coverSet(statesNeeded, stations) {
    const finalStaions = new Set();

    while (statesNeeded.size != 0) {
        let bestStaton = null;
        let statesCovered = new Set();

        for (const [station, statesForStatoin] of Object.entries(stations)) {
            const covered = intersection(statesNeeded, statesForStatoin);

            if (covered.size > statesCovered.size) {
                bestStaton = station;
                statesCovered = covered;
            }
        }

        finalStaions.add(bestStaton);

        statesNeeded = difference(statesNeeded, statesCovered);
    }

    return finalStaions;
}

console.log(coverSet(statesNeeded, stations));
